src/GraphTask.Plugins/EdgeColoringPlugin/EdgeColoringPlugin.cs

```diff
@@ -0,0 +1,145 @@
+    using System.Windows;
+    using QuickGraph.Algorithms;
+    [Extension]
+    public class VertexColoringPlugin : IAlgorithm
+    {
+        private readonly GraphArea _graphArea;
+        private readonly GraphXZoomControl _zoomControl;
+        private bool _hasStarted;
+        private bool _hasFinished;
+        private List<Tuple<GraphXVertex, GraphXVertex, int>> states;
+        private EdgeColoringAlgorithm<GraphXVertex, TaggedEdge<GraphXVertex, int>> coloringAlgorithm;
+        private UndirectedGraph<GraphXVertex, TaggedEdge<GraphXVertex, int>> graphWithColoredVertices;
+        private BiGraph bigraph;
+        private int currStateNumber;
+        private String[] colors;
+
+        public VertexColoringPlugin()
+        {
+            _graphArea = new GraphArea();
+            _zoomControl = new GraphXZoomControl { Content = _graphArea };
+            Output.Controls.Add(new ElementHost { Dock = DockStyle.Fill, Child = _zoomControl });
+        }
+
+        public string Name => "Edge Coloring Plugin";
+        public string Author => "Alex Rabochiy";
+        public string Description => "This plugin demonstrates how Edge Coloring works.\n";
+
+        public Panel Options { get; } = new Panel { Dock = DockStyle.Fill };
+        public Panel Output { get; } = new Panel { Dock = DockStyle.Fill };
+
+        public bool CanGoBack => currStateNumber != 0 && states.Count != 0;
+        public bool CanGoFurther => _hasStarted && !_hasFinished;
+
+        public void Run(string dotSource)
+        {
+            var input = ReadInputGraph(dotSource);
+            if (!input.Vertices.Any())
+            {
+                MessageBox.Show("Graph is empty.");
+                return;
+            }
+            colors = ColorGenerator.genColors(input.EdgeCount);
+            coloringAlgorithm = new EdgeColoringAlgorithm<GraphXVertex, TaggedEdge<GraphXVertex, int>>();
+
+            states = new List<Tuple<GraphXVertex, GraphXVertex, int>>();
+
+            coloringAlgorithm.Colored += OnColourEdge;
+            currStateNumber = -1;
+
+            graphWithColoredVertices = coloringAlgorithm.Compute(input);
+
+            _graphArea.LogicCore.Graph = bigraph;
+            _graphArea.GenerateGraph();
+            _zoomControl.ZoomToFill();
+
+            _hasStarted = true;
+            _hasFinished = false;
+        }
+
+        private void OnColourEdge(object sender, Tuple<GraphXVertex, GraphXVertex, int> args)
+        {
+            states.Add(args);
+        }
+
+        private void HighlightTraversal()
+        {
+            var currState = states.ElementAt(currStateNumber);
+            GraphXVertex source = currState.Item1;
+            GraphXVertex target = currState.Item2;
+            GraphXTaggedEdge<GraphXVertex, int> edge = null;
+            foreach (var key in _graphArea.EdgesList.Keys)
+            {
+                if (key.Source.Text.Equals(source.Text) && key.Target.Text.Equals(target.Text))
```

Comment:
Не уверен. Странно, что у Вас разные вершины получаются. Вроде должны быть
одни и те же, а различаются только рёбра.

Regards,
Semyon Grigorev.

2016-06-14 13:37 GMT+03:00 Alexey Rabochy <notifications@github.com>:

> In src/GraphTask.Plugins/EdgeColoringPlugin/EdgeColoringPlugin.cs
> <https://github.com/YaccConstructor/QuickGraph/pull/137#discussion_r66946452>
> :
>
> > +        }
> > +
> > +        private void OnColourEdge(object sender, Tuple<GraphXVertex, GraphXVertex, int> args)
> > +        {
> > +            states.Add(args);
> > +        }
> > +
> > +        private void HighlightTraversal()
> > +        {
> > +            var currState = states.ElementAt(currStateNumber);
> > +            GraphXVertex source = currState.Item1;
> > +            GraphXVertex target = currState.Item2;
> > +            GraphXTaggedEdge<GraphXVertex, int> edge = null;
> > +            foreach (var key in _graphArea.EdgesList.Keys)
> > +            {
> > +                if (key.Source.Text.Equals(source.Text) && key.Target.Text.Equals(target.Text))
>
> Допустимо ли тогда сравнивать ID вершин?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/YaccConstructor/QuickGraph/pull/137/files/02d588dfedbf5dd4efd2cfcf17ccb985065ebb84#r66946452>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAGhQBcqmTk3PNHerMmT_g_MtWO1Gtt_ks5qLoRrgaJpZM4Iz0x2>
> .
>